kanidm_unix_int/nss_kanidm/src/lib.rs

```diff
@@ -10,162 +10,15 @@
-    }
-}
-
-struct KanidmGroup;
-libnss_group_hooks!(kanidm, KanidmGroup);
-
-impl GroupHooks for KanidmGroup {
-    fn get_all_entries() -> Response<Vec<Group>> {
-        let cfg =
-            match KanidmUnixdConfig::new().read_options_from_optional_config(DEFAULT_CONFIG_PATH) {
-                Ok(c) => c,
-                Err(_) => {
-                    return Response::Unavail;
-                }
-            };
-        let req = ClientRequest::NssGroups;
-        call_daemon_blocking(cfg.sock_path.as_str(), &req, cfg.unix_sock_timeout)
-            .map(|r| match r {
-                ClientResponse::NssGroups(l) => l.into_iter().map(group_from_nssgroup).collect(),
-                _ => Vec::new(),
-            })
-            .map(Response::Success)
-            .unwrap_or_else(|_| Response::Success(vec![]))
-    }
-
-    fn get_entry_by_gid(gid: libc::gid_t) -> Response<Group> {
-        let cfg =
-            match KanidmUnixdConfig::new().read_options_from_optional_config(DEFAULT_CONFIG_PATH) {
-                Ok(c) => c,
-                Err(_) => {
-                    return Response::Unavail;
-                }
-            };
-        let req = ClientRequest::NssGroupByGid(gid);
-        call_daemon_blocking(cfg.sock_path.as_str(), &req, cfg.unix_sock_timeout)
-            .map(|r| match r {
-                ClientResponse::NssGroup(opt) => opt
-                    .map(group_from_nssgroup)
-                    .map(Response::Success)
-                    .unwrap_or_else(|| Response::NotFound),
-                _ => Response::NotFound,
-            })
-            .unwrap_or_else(|_| Response::NotFound)
-    }
-
-    fn get_entry_by_name(name: String) -> Response<Group> {
-        let cfg =
-            match KanidmUnixdConfig::new().read_options_from_optional_config(DEFAULT_CONFIG_PATH) {
-                Ok(c) => c,
-                Err(_) => {
-                    return Response::Unavail;
-                }
-            };
-        let req = ClientRequest::NssGroupByName(name);
-        call_daemon_blocking(cfg.sock_path.as_str(), &req, cfg.unix_sock_timeout)
-            .map(|r| match r {
-                ClientResponse::NssGroup(opt) => opt
-                    .map(group_from_nssgroup)
-                    .map(Response::Success)
-                    .unwrap_or_else(|| Response::NotFound),
-                _ => Response::NotFound,
-            })
-            .unwrap_or_else(|_| Response::NotFound)
-    }
-}
-
-fn passwd_from_nssuser(nu: NssUser) -> Passwd {
-    Passwd {
-        name: nu.name,
-        gecos: nu.gecos,
-        passwd: "x".to_string(),
-        uid: nu.gid,
-        gid: nu.gid,
-        dir: nu.homedir,
-        shell: nu.shell,
-    }
-}
-
-fn group_from_nssgroup(ng: NssGroup) -> Group {
-    Group {
-        name: ng.name,
-        passwd: "x".to_string(),
-        gid: ng.gid,
-        members: ng.members,
-    }
-}
+#[cfg(target_family = "unix")]
```

Comment:
Do we actually need this pub use? Given the way that the nsswitch module works, it doesn't matter what file the implementation is in because of howe the symbols are exported.

Reply:
:shrug: I'm not sure how any of it works TBH, I assumed it was based on being exported from the root of the module somehow?

Reply:
Yes, but not in the way you might think. pub use is a rust-specific construct that is only understood by the rust compiler and has no impact on the resulting symbols in the shared library.

The actual magic is in the "pam_hooks" macro, which defines a c extern / no_mangle which is a function name that is NOT dead code eliminated, and also has a static symbol name in the dll/so. The location of it within the rust source tree doesn't matter, because it always exports that symbol name into the resulting object.

So the pub use here isn't needed, because we don't have any consuming rust code that links into this, and the pam_hooks call is exporting the symbol names correctly in a way that is indepedent to the rust source layout.

Does that help,

Reply:
Sounds good to me, I'll take it out and we'll see how it goes 😄

Reply:
As discussed offline - william was wrong!

Reply:
Amazing! It does happen sometimes. ❤️